sdk/src/logs/batch_log_processor.cc

```diff
@@ -0,0 +1,222 @@
+    {
+      // Since this export was the result of a force flush, signal the
+      // main thread that the worker thread has been notified
+      is_force_flush_.store(false);
+    }
+    else
+    {
+      // If the buffer was empty during the entire `timeout` time interval,
+      // go back to waiting. If this was a spurious wake-up, we export only if
+      // `buffer_` is not empty. This is acceptable because batching is a best
+      // mechanism effort here.
+      if (buffer_.empty() == true)
+      {
+        continue;
+      }
+    }
+
+    auto start = std::chrono::steady_clock::now();
+
+    Export(was_force_flush_called);
+
+    auto end      = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+    // Subtract the duration of this export call from the next `timeout`.
+    timeout = schedule_delay_millis_ - duration;
+  }
+}
+
+void BatchLogProcessor::Export(const bool was_force_flush_called)
+{
+  std::vector<std::unique_ptr<Recordable>> records_arr;
+
+  size_t num_records_to_export;
+
+  if (was_force_flush_called == true)
+  {
+    num_records_to_export = buffer_.size();
+  }
+  else
+  {
+    num_records_to_export =
+        buffer_.size() >= max_export_batch_size_ ? max_export_batch_size_ : buffer_.size();
+  }
+
+  buffer_.Consume(
+      num_records_to_export, [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+          std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+          ptr.Swap(swap_ptr);
+          records_arr.push_back(std::unique_ptr<Recordable>(swap_ptr.release()));
+          return true;
+        });
+      });
+
+  if (exporter_->Export(records_arr) != ExportResult::kSuccess)
+  {
+    // Indicate Error: "[Batch Log Processor]: Failed to export a batch"
+  }
+
+  // Notify the main thread in case this export was the result of a force flush.
+  if (was_force_flush_called == true)
+  {
+    is_force_flush_notified_.store(true);
+    // Notifies the thread
+    force_flush_cv_.notify_one();
+  }
+}
+
+void BatchLogProcessor::DrainQueue()
+{
+  while (buffer_.empty() == false)
+  {
+    Export(false);
+  }
+}
+
+// Note: Timeout functionality is currently not implemented
+bool BatchLogProcessor::Shutdown(std::chrono::microseconds timeout) noexcept
+{
+  // Atomically checking whether value of is_shutdown_ is false
+  // and setting it to true
+  if (is_shutdown_.exchange(true) == false)
+  {
+    // notifies worker thread that shutdown has been called
+    cv_.notify_one();
+    // wait until worker thread completes current export
+    worker_thread_.join();
+    // calls the exporter to shutdown
+    if (exporter_ != nullptr)
+    {
+      return exporter_->Shutdown();
+    }
+    return true;
+  }
+  return false;
+}
+
+BatchLogProcessor::~BatchLogProcessor()
+{
+  if (is_shutdown_.load() == false)
```

Comment:
Nit: Just call shutdown here.  You're guarding calling shutdown twice anyway.

Reply:
Yeah for sure, will make the change!